### PR TITLE
allow importing add-on code from an addons namespace

### LINKED ISSUE #14359: Add package for addons to store plugin independent code
### Is your feature request related to a problem? Please describe.
An addon can contain appModules, brailleDisplayDrivers, synthDrivers, globalPlugins and visionEnhancementProviders. However, there is no way for addons to store code that has to be shared between packages. Imagine a case where a constant has to be defined that has to be available in both an appModule and a globalPlugin. I'm pretty sure @josephsl might know of other cases like this.
In my case, I want to store a base class subclassing driverHandler.Driver that will be a base of both a synthDriver and brailleDisplayDriver.

### Describe the solution you'd like
I'd like to suggest the possibility for addons to provide code for the following NVDA packages:

* comInterfaces: for addons requiring com interface definitions that can be resolved with comTypes automatically (i.e. without doing a magic import trick to make the comInterface available to comtypes)
* utils: package that can hold classes and modules that ought to be shared between plugins in the addon

Cc @josephsl and other addon contributors

### PR DESCRIPTION
**Pull request by LeonarddeR**
Cc @josephsl 

### Link to issue number:
Closes #14359

### Summary of the issue:
It is currently only possible to import add-on code from the standard NVDA plugin paths, like appModules or globalPlugins. Therefore, there is no elegant way add-ons can store code that has to be shared between appModules and globalPlugins, for example.

### Description of user facing changes
@feerrenrut suggested the following approach in https://github.com/nvaccess/nvda/issues/14359#issuecomment-1314699202:

> I'd prefer that addon code isn't mixed in with the standard namespaces. Instead, a solution that maintained the separation of the add-on.
> 
> When importing shared code from within the addon this should be achieved in a way that avoids namespace collisions: Given an addon called `myAddon` you might import code from a `lib` directory within the add-on with `from myAddon.lib import comInterfaces`
> 
> * This would help to prevent add-ons from inadvertently depending on each other.
> * We don't want to end up in a situation where the addition of a new file in NVDA could break add-ons because of namespace pollution.

The final solution is slightly different in that it adds an `addons` part in the naming scheme: Given an addon called `myAddon` you can import code from a `lib` directory within the add-on with `from addons.myAddon.lib import comInterfaces.

### Description of development approach
This solution depends heavily on importlib. It generates a blank module spec based on a module name and optional path, and constructs a module from that spec. The module is inserted in sys.modules. The addons themselves are simulated [Native Namespace Packages](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages). Long story short, these are python packages that don't require an `__init__.py` file.

### Testing strategy:
Created an addon called `example`. In the addon folder, added a folder called `lib` and in there a module called `test.py`. Successfully imported the module with `from addons.example.lib import test`

### Known issues with pull request:
None known

### Change log entries:
For Developers
- Add-ons can now contain python code anywhere in their directory structure. The contents of an add-on folder are exposed under the addons namespace. For example, a test module in an add-on called example can be imported with "from addons.example import test". (#14359)

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

File: source/addonHandler/packaging.py [hunk 1]
```diff
@@ -7,11 +7,15 @@
 """
 
 import os.path
-from typing import Optional
+from typing import Optional, List
```
Comment (seanbudd) [inline source/addonHandler/packaging.py:10]:
```suggestion
from typing import (
	List,
	Optional,
)
```

**Review by seanbudd — APPROVED** [review]
Generally looks good to me

**Comment by feerrenrut** [comment]
@leonardder The implementation so far looks good.

This might be a good candidate for an automated test. It's functionality that would be easy for NVDA core devs to forget about. It would need to be a system test, perhaps two simple add-ons, one that shares the code from the other and does something like announces the "hello world" message sourced from the other add-on.

One other aspect to consider here is managing dependencies between add-ons. This will become a new problem, add-ons that only work when others are installed. Users will need a way to find out what dependencies to install and receive warnings when the dependencies aren't met.

**Comment by leonardder** [comment]
> One other aspect to consider here is managing dependencies between add-ons. This will become a new problem, add-ons that only work when others are installed. Users will need a way to find out what dependencies to install and receive warnings when the dependencies aren't met.

I think that interdependent add-ons are out of scope for this solution. Note that this possibility is already available. If add-on 1 bundles a globalPlugin and add-on 2 bundles a synthDriver, it is possible to do `from globalPlugins import addonOnePlugin` from add-on two and vise versa. The primary goal with this pr is that an add-on can bundle plugin independent code that can be used in several plugins in the add-on itself.

File: devDocs/developerGuide.t2t [hunk 2]
```diff
@@ -845,6 +846,20 @@ NVDA will look for and execute an onUninstall function in installTasks.py when N
 After this function completes, the add-on's directory will automatically be removed.
 As this happens on NVDA startup before other components are initialized, this function cannot request input from the user.
 
+++ Optional plugin independent code ++
+In NVDA 2023.1 and above, it is possible to bundle plugin independent code with an add-on.
+This can be used to bundle base classes, constants, etc. that have to be available to multiple plugins in the add-on.
+This is especially useful if your add-on bundles an appModule and a globalPlugin and both plugins need to access the same class bundled with the add-on, for example.
```
Comment (CyrilleB79) [inline devDocs/developerGuide.t2t:852]: A detail: "especially" and "for example" do not seem so compatible in this phrasing. So only use one of them depending on what you want to say.

```suggestion
For example, this is useful if your add-on bundles an appModule and a globalPlugin and both plugins need to access the same class bundled with the add-on.
```

or:

```suggestion
This is especially useful if your add-on bundles an appModule and a globalPlugin and both plugins need to access the same class bundled with the add-on.
```

**Comment by leonardder** [comment]
Looking at the system tests library stuff, it all seems a bit hacky to me. There is a special function that constructs a scratchpad from several modules. Why was this made like this? Should we consider making this an add-on instead?

**Comment by feerrenrut** [comment]
> The primary goal with this pr is that an add-on can bundle plugin independent code that can be used in several plugins in the add-on itself.

While it was possible before, this PR is introducing interdependence between add-ons as a first-class use-case. Regardless of whether it's in scope of this PR or not, this is a problem that will need to be solved either in the add-ons or through NVDA add-on management. It's a tricky problem, the phrase "dependency hell" exists for a reason. Consider:
- `Addon-A` needs `version 1` of `Addon-base`
- `Addon-B` needs `version 2` of `Addon-base`

It's very tempting to say this isn't a path we want to go down, add-ons should just duplicate dependencies they need.
Instead, it might be better to look at how the different parts of an add-on (appModules, brailleDisplayDrivers, synthDrivers, globalPlugins and visionEnhancementProviders) can share code internally, but prevent leaking this to other add-ons.
I'd recommend a dependency injection approach:
- A `lib` package in the addon contain code to be shared within this add-on.
- NVDA imports this and then injects it into each plugin type via a `receiveLib()` function with a default implementation.
- I'm not sure what the most appropriate thing to inject would be. Can it be the module? If not, a static class might be more appropriate, or perhaps a callable with the result implementation defined (I.E. `receiveLib(getLib: Callable[(), Any])`)?

>There is a special function that constructs a scratchpad from several modules. Why was this made like this? Should we consider making this an add-on instead?

Off topic for this PR, but happy for alternatives to be considered in a new issue/PR. Using the scratchpad approach was taken because it was considered simpler while fundamentally providing the same outcome. This approach means there is no need for an add-on packaging step and no add-on install step. While working on the tests, and potentially changing libraries, the files need to be redeployed to ensure the tests run with the latest version. This happens for every 'start NVDA'.

**Comment by leonardder** [comment]
> It's very tempting to say this isn't a path we want to go down, add-ons should just duplicate dependencies they need. Instead, it might be better to look at how the different parts of an add-on (appModules, brailleDisplayDrivers, synthDrivers, globalPlugins and visionEnhancementProviders) can share code internally, but prevent leaking this to other add-ons. I'd recommend a dependency injection approach:
> 
> * A `lib` package in the addon contain code to be shared within this add-on.
> * NVDA imports this and then injects it into each plugin type via a `receiveLib()` function with a default implementation.
> * I'm not sure what the most appropriate thing to inject would be. Can it be the module? If not, a static class might be more appropriate, or perhaps a callable with the result implementation defined (I.E. `receiveLib(getLib: Callable[(), Any])`)?

Interesting idea. I will research this further. A similar approach is used for bundling installTasks with an add-on.

**Comment by leonardder** [comment]
I discovered a pretty straight forward method to do this already.
An add-on can call `addonHandler.getCodeAddon()` and then `loadModule` on the add-on to get a module from its directory. I think that fully suffices the use case I'm proposing. It makes sense to document it tough

**Pull request closed without merging**